### ISSUE TITLE
[pd~]: subprocess dies immediately

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1953]
In the [pd~] help patch, when I try to open the subprocess with `[pd~ start pd~-subprocess.pd(` , the subprocess GUI shows for a brief moment and then the subprocess dies. The Pd console prints `pd~: subprocess exited` and in the terminal I see the following output:
```
pd-extern: EOF on input
pdsend errorname: >>error writing "sock0000000003FC3B00": connection reset by peer<<
pdsend errorname: >>error writing "sock0000000003FC3B00": connection reset by peer<<
pdsend errorname: >>error writing "sock0000000003FC3B00": connection reset by peer<<
pdsend errorname: >>error writing "sock0000000003FC3B00": connection reset by peer<<
pdsend errorname: >>error writing "sock0000000003FC3B00": connection reset by peer<<
```
This happens on Pd 0.53-2, Windows 10. Can anyone confirm this?

BTW, starting the subprocess without GUI, i.e. `[pd~ start -nogui pd~-subprocess.pd(`, works fine.

**Comment by Lucarda** [comment]
Can't reproduce on Win11

|pd-0.53-1|pd-0.53-2| |
|:--|:--|:--|
|ASIO|ASIO|works fine|
|MMIO|MMIO|works fine|

**Comment by Spacechild1** [comment]
I got a bit further:

1. It only crashes when the parent Pd process is started as `pd.com` (either explicitly via the command line, or implicitly via the GUI), but it works fine if started as `pd.exe`

2. Pd 0.53-0 is the first version to have this issue

@Lucarda Can you confirm 1.?

**Comment by Spacechild1** [comment]
It gets even stranger: When I run `pd.exe` by double-clicking, it works, but when I run it on the command line, the subprocess dies and I get the following error messages in the terminal (which I do not get with `pd.com`):
```
pd-extern: EOF on input
pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<
pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<
pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<
pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<
pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<
```

**Comment by Spacechild1** [comment]
The regression must have been introduced between Mar 28, 2022 (Pd 0.52-2) and Oct 25, 2022 (Pd 0.53). The last code change to `[pd~]` itself happened before (https://github.com/pure-data/pure-data/commit/23a0b864d4fd5f64e854e92eb5d382e0c2ef851e)

**Comment by Lucarda** [comment]
> @Lucarda Can you confirm 1.?

@Spacechild1 I can't get the crash. I'm starting Pd-0.53-2 from the Win11 terminal (cmd) via: 
`pd.com`, 
`pd.exe`
 or `wish86.exe ..\tcl\pd-gui.tcl`.
 Also I tested all with both ASIO and MMIO.

**Comment by Lucarda** [comment]
I just did the same tests as above but on **Win10** (same hardware) and could not get the crash either.

**Comment by Spacechild1** [comment]
Now that is *very* strange.

**Comment by Lucarda** [comment]
indeed. I think you know the error is coming from https://github.com/pure-data/pure-data/blob/3ae0388c2ba354e2b5d818c1422b7484c4a491bf/tcl/pd_connect.tcl#L63 but i have no clue on whats going on.

**Comment by Spacechild1** [comment]
Here's the culprit: c4ae1bd8df436de15c0a80f303ae33b5f3569088. The previous commit (fbac4bc2e642b71656f47a40aa866e69e6cd91e0) still works. @umlaeute Any ideas?

There is something weird about the error message mentioned earlier:

> pdsend errorname: >>error writing "sock00000000040732B0": connection reset by peer<<

This comes from the GUI, but why do I see it in my terminal when running `pd.exe`? One would expect that the output of a child process is *not* automatically redirected to that of the parent process... But on Windows this can happen depending on how the subprocess is created (https://learn.microsoft.com/en-us/windows/console/creation-of-a-console).

Here's an idea: maybe the GUI prints something to *stdout* which gets accidentally redirected to the subprocess; this would trip up `[pd~]` because it consumes the stdout of the subprocess via a pipe and expects valid Pd messages.

This would also explain why it works with double clicking `pd.exe` (= no console), but neither with double clicking `pd.com` nor with running `pd.exe` or `pd.com` from the terminal (= console) since they all create a console.

If my assessment is correct, we would need to make sure that the GUI process is created in a way that it does not forward its stdout to the core.

**Comment by Lucarda** [comment]
> This would also explain why it works with double clicking pd.exe (= no console), but neither with double clicking pd.com nor with running pd.exe or pd.com from the terminal (= console) since they all create/inherit a console.

For me "running pd.exe" from a terminal returns automatically and will not print anything later on. The only one that does not return and prints stuff to the terminal is `pd.com`.

**Comment by Lucarda** [comment]
@Spacechild1 i think I cough you. You are using Msys2 terminal :)

I can reproduce the crash on my Msys2 terminal but not on Windows terminals.

**Comment by Spacechild1** [comment]
> I can reproduce the crash on my Msys2 terminal but not on Windows terminals.

Phew. I thought I was hallicinating. I guess different terminals use different options in `CreateProcess` when starting new applications.

**Comment by Spacechild1** [comment]
I still don't really know what the problem is, but I finally figured out the right options/flags to make the problem disappear :-) Will make a PR soon for others to test.

**Comment by Spacechild1** [comment]
Should be fixed with 0cbda8e969e529fe71419d6d92b8ad4f3a1f2936 on the develop branch. @Lucarda If you have time, please double-check!

**Comment by Lucarda** [comment]
@Spacechild1 it does work for me both on Windows and Msys2 terminals.

When running `./pd` or `./wish85.exe ../tcl/pd-gui.tcl` on Msys2 terminal I get: 

```
pd-extern: EOF on input
```
when I stop the subprocess but I think this is Ok and expected.

--
Tested on Win11 and MMIO.

**Comment by Spacechild1** [comment]
Thanks!

> when I stop the subprocess but I think this is Ok and expected.

Yes, that is just a message from the subprocess before it ends, it does not (necessarily) indicate an error. (I'm working on a branch with some `[pd~]` improvements, and one part is about better error messsages :-)

**Comment by Spacechild1** [comment]
Closed via 0cbda8e969e529fe71419d6d92b8ad4f3a1f2936
